janus_server/src/bin/aggregation_job_driver.rs

```diff
@@ -0,0 +1,1422 @@
+                    if let PrepareTransition::Finish(out_share) = leader_transition {
+                        report_aggregation.state = ReportAggregationState::Finished(out_share);
+                    } else {
+                        error!(task_id = %task.id, aggregation_job_id = %aggregation_job.aggregation_job_id,
+                            report_nonce = %report_aggregation.nonce, leader_transition = ?leader_transition, "Helper finished but leader did not");
+                        report_aggregation.state = ReportAggregationState::Invalid;
+                    }
+                }
+
+                TransitionTypeSpecificData::Failed { error } => {
+                    // If the helper failed, we move to FAILED immediately.
+                    // TODO(brandon): is it correct to just record the transition error that the helper reports?
+                    error!(task_id = %task.id, aggregation_job_id = %aggregation_job.aggregation_job_id,
+                        report_nonce = %report_aggregation.nonce, helper_err = ?error, "Helper couldn't step report aggregation");
+                    report_aggregation.state = ReportAggregationState::Failed(error);
+                }
+            }
+            report_aggregations_to_write.push(report_aggregation);
+        }
+
+        // Determine if we've finished the aggregation job (i.e. if all report aggregations are in
+        // a terminal state), then write everything back to storage.
+        let aggregation_job_id = aggregation_job.aggregation_job_id;
+        let aggregation_job_is_finished = report_aggregations_to_write
+            .iter()
+            .all(|ra| !matches!(ra.state, ReportAggregationState::Waiting(_, _)));
+        let aggregation_job_to_write = if aggregation_job_is_finished {
+            let mut aggregation_job = aggregation_job;
+            aggregation_job.state = AggregationJobState::Finished;
+            Some(aggregation_job)
+        } else {
+            None
+        };
+        let report_aggregations_to_write = Arc::new(report_aggregations_to_write);
+        let aggregation_job_to_write = Arc::new(aggregation_job_to_write);
+        self.datastore
+            .run_tx(|tx| {
+                let (report_aggregations_to_write, aggregation_job_to_write) = (
+                    Arc::clone(&report_aggregations_to_write),
+                    Arc::clone(&aggregation_job_to_write),
+                );
+                Box::pin(async move {
+                    let report_aggregations_future =
+                        try_join_all(report_aggregations_to_write.iter().map(
+                            |report_aggregation| tx.update_report_aggregation(report_aggregation),
+                        ));
+                    let aggregation_job_future = try_join_all(
+                        aggregation_job_to_write
+                            .iter()
+                            .map(|aggregation_job| tx.update_aggregation_job(aggregation_job)),
```

Comment:
I agree; I was considering changing things in this PR, but it's already way too large. I'd like to follow up on this in another PR unless folks object strongly. But let's chat about the design a bit:

I think re-reading everything would work, but it might be too much work -- we might be reading & checking hundreds of report aggregations, for example. I was instead thinking of having the "acquire" operation return a "lease token" that would need to be piped through to the "release" operation, and checked on releasing. The token would be comprised of at least the timestamp, and probably a few bytes of randomness; the check would consist of making sure the token matches up to the token recorded in the DB, and that the end-time has not passed. [I'm thought about clever ways to replace the "randomness" instead with something like the "hash of the data covered by the lease", but that would be pretty hard to get right -- e.g. on updates, we might not always want to reread all of the data to produce the hash.]

(If we ever want to break our helper's holding a TX open while processing aggregate init/continue requests, we could use a similar technique; we wouldn't want a timed lease, but we could keep an opaque value around to detect & retry on concurrent updates.)

Reply:
I like the lease token idea. I happened upon a good writeup of how to solve that a few weeks ago: https://martin.kleppmann.com/2016/02/08/how-to-do-distributed-locking.html